src/client/datascience/notebook/diagnosticsProvider.ts

```diff
@@ -0,0 +1,234 @@
+            if (d.message !== pipMessage && d.message !== condaMessage) {
+                return;
+            }
+            const isPip = d.message === pipMessage;
+            const installer = isPip ? 'pip' : 'conda';
+            const codeAction = new CodeAction(
+                DataScience.replacePipCondaInstallCodeAction().format(installer),
+                CodeActionKind.QuickFix
+            );
+            codeAction.isPreferred = true;
+            codeAction.diagnostics = [d];
+            const edit = new WorkspaceEdit();
+            edit.replace(
+                document.uri,
+                new Range(d.range.start, new Position(d.range.start.line, d.range.start.character + 1)),
+                '%'
+            );
+            codeAction.edit = edit;
+            codeActions.push(codeAction);
+        });
+        return codeActions;
+    }
+    private analyzeNotebook(notebook: NotebookDocument): void {
+        if (notebook.notebookType !== JupyterNotebookView) {
+            return;
+        }
+        // Process just the first 100 cells to avoid blocking the UI.
+        notebook.getCells().forEach((cell, i) => (i < 100 ? this.queueCellForProcessing(cell) : undefined));
+    }
+
+    private queueCellForProcessing(cell: NotebookCell): void {
+        this.cellsToProces.add(cell);
+        this.analyzeNotebookCells();
+    }
+    private analyzeNotebookCells() {
+        if (this.cellsToProces.size === 0) {
+            return;
+        }
+        const cell = this.cellsToProces.values().next().value;
+        this.cellsToProces.delete(cell);
+        this.analyzeNotebookCell(cell);
+        // Schedule processing of next cell (this way we dont chew CPU and block the UI).
+        setTimeout(() => this.analyzeNotebookCells(), 0);
+    }
+    private analyzeNotebookCell(cell: NotebookCell) {
+        if (
+            cell.kind !== NotebookCellKind.Code ||
+            cell.document.languageId !== PYTHON_LANGUAGE ||
+            cell.notebook.isClosed ||
+            cell.document.isClosed
+        ) {
+            return;
+        }
+        // If we've already process this same cell, and the version is the same, then we don't need to do anything.
+        if (this.notebooksProcessed.get(cell.notebook)?.get(cell.document.uri.toString()) === cell.document.version) {
+            return;
+        }
+
+        this.problems.delete(cell.document.uri);
+        const cellsUrisWithProblems = this.notebooksProcessed.get(cell.notebook) || new Map<CellUri, CellVersion>();
+        cellsUrisWithProblems.set(cell.document.uri.toString(), cell.document.version);
+        this.notebooksProcessed.set(cell.notebook, cellsUrisWithProblems);
+
+        // For perf reasons, process just the first 50 lines.
+        for (let i = 0; i < Math.min(cell.document.lineCount, 50); i++) {
+            const line = cell.document.lineAt(i);
+            const text = line.text;
+            if (text.trim().startsWith('!pip install')) {
```

Comment:
These can also be inside an if condition so you might want to use a regex instead.

Reply:
Never mind, missed the trim.

Reply:
Even if inside if condition, this would be in a separate line